src/main/java/org/jboss/modules/JDKSpecific.java

```diff
@@ -185,6 +189,36 @@ static void addInternalPackages(final List<String> list) {
         list.add("jdk.internal.reflect.");
     }
 
+    static <T> Iterable<T> findServices(final Class<T> serviceType, final Predicate<Class<?>> filter, final ClassLoader classLoader) {
+        final Iterator<T> delegate = ServiceLoader.load(serviceType, classLoader).iterator();
+        return new Iterable<T>() {
+            public Iterator<T> iterator() {
+                return new Iterator<T>() {
+                    T next = null;
+
+                    public boolean hasNext() {
+                        T next;
+                        while (delegate.hasNext()) {
+                            next = delegate.next();
+                            if (filter.test(next.getClass())) {
+                                this.next = next;
+                                return true;
+                            }
+                        }
+                        return false;
+                    }
+
+                    public T next() {
+                        if (this.next == null) throw new NoSuchElementException();
```

Comment:
This should call `hasNext()`; it's allowed to call `next()` without calling `hasNext()`.

Reply:
Right!